File: NOVEMBER-2025/28-11-2025/2872.go
func maxKDivisibleComponents(n int, edges [][]int, values []int, k int) int {
    g := make([][]int, n)
    for _, e := range edges {
        a, b := e[0], e[1]
        g[a] = append(g[a], b)
        g[b] = append(g[b], a)
    }

    ans := 0

    var dfs func(u, p int) int64
    dfs = func(u, p int) int64 {
        sum := int64(values[u])
        for _, v := range g[u] {
            if v == p {
                continue
            }
            sum += dfs(v, u)
        }
        if sum%int64(k) == 0 {
            ans++
            return 0
        }
        return sum
    }

    dfs(0, -1)
    return ans
}

File: NOVEMBER-2025/28-11-2025/2872.py
class Solution:
    def maxKDivisibleComponents(self, n: int, edges: List[List[int]], values: List[int], k: int) -> int:
        g = [[] for _ in range(n)]
        for a, b in edges:
            g[a].append(b)
            g[b].append(a)

        self.ans = 0

        def dfs(u, p):
            s = values[u]
            for v in g[u]:
                if v == p:
                    continue
                s += dfs(v, u)
            if s % k == 0:
                self.ans += 1
                return 0
            return s

        dfs(0, -1)
        return self.ans




class Solution(object):
    def maxKDivisibleComponents(self, n, edges, values, k):
        g = [[] for _ in range(n)]
        for a, b in edges:
            g[a].append(b)
            g[b].append(a)

        self.ans = 0

        def dfs(u, p):
            s = values[u]
            for v in g[u]:
                if v == p:
                    continue
                s += dfs(v, u)
            if s % k == 0:
                self.ans += 1
                return 0
            return s

        dfs(0, -1)
        return self.ans